### PR TITLE
Offramp from RRD if veteran has pending EP 020

### PR DESCRIPTION
**Pull request by nanotone**
## Description of change
This PR changes the DisabilityCompensationJob class to off-ramp in the event that the claim is attached to any pending EPs.

Additionally, there's a fix related to the use of the[ `return` vs `next` keywords](https://github.com/department-of-veterans-affairs/vets-api/pull/9344/files#r826368786).

## Original issue(s)

https://www.notion.so/Set-up-monitoring-of-off-ramped-EP400-claims-2269302a9a934743b4126f717cf23804

### CODE REVIEW

**Comment by allthesignals** [comment]
In order to re-record VCRs for this, I'll need to setup EVSS access, and I’m struggling to find documentation on that (aside from a README that tells me to ask a PM for it). I'll poke around and see if I can't get that setup!

File: spec/jobs/rapid_ready_for_decision/disability_compensation_job_spec.rb [hunk 1]
```diff
@@ -174,10 +194,23 @@
...
+            expect do
+              subject.perform_async(submission.id)
+            end.to raise_error(ArgumentError, 'no ICN passed in for LH API request.')
+          end
+        end
+      end
+    end
+
+    context 'when there are pending claims that cause EP 400 errors' do
+      it 'off ramps onto the regular process' do
```
Comment (yoomlam) [inline spec/jobs/rapid_ready_for_decision/disability_compensation_job_spec.rb:208]:
```suggestion
      it 'off-ramps to the non-RRD process' do
```

File: app/workers/rapid_ready_for_decision/disability_compensation_job.rb [hunk 2]
```diff
@@ -27,6 +27,8 @@ def perform(form526_submission_id)
 
       begin
         with_tracking(self.class.name, form526_submission.saved_claim_id, form526_submission_id) do
+          return if pending_eps?(form526_submission.auth_headers)
```
Comment (yoomlam) [inline app/workers/rapid_ready_for_decision/disability_compensation_job.rb:30]: I believe Yang found a possible bug: https://github.com/department-of-veterans-affairs/vets-api/pull/9344/files#r826368786

File: spec/jobs/rapid_ready_for_decision/disability_compensation_job_spec.rb [hunk 3]
```diff
@@ -174,10 +194,23 @@
...
+        VCR.use_cassette('evss/claims/claims') do
+          Sidekiq::Testing.inline! do
+            expect(RapidReadyForDecision::HypertensionMedicationRequestData).not_to receive(:new)
```
Comment (yoomlam) [inline spec/jobs/rapid_ready_for_decision/disability_compensation_job_spec.rb:211]: How about checking at a higher/earlier level:
```suggestion
            expect(Lighthouse::VeteransHealth::Client).not_to receive(:new)
```

In addition, you could check that `pending_eps?(form526_submission.auth_headers)` returns true.

File: spec/jobs/rapid_ready_for_decision/disability_compensation_job_spec.rb [hunk 4]
```diff
@@ -174,10 +194,23 @@
...
+          Sidekiq::Testing.inline! do
+            expect do
+              subject.perform_async(submission.id)
+            end.to raise_error(ArgumentError, 'no ICN passed in for LH API request.')
+          end
+        end
+      end
+    end
+
+    context 'when there are pending claims that cause EP 400 errors' do
```
Comment (yoomlam) [inline spec/jobs/rapid_ready_for_decision/disability_compensation_job_spec.rb:207]: I believe you mean:
```suggestion
    context 'when there are pending claims, which cause EP 400 errors' do
```
(since this test doesn't actually check for EP 400 errors; it checks for pending claims).

File: app/workers/rapid_ready_for_decision/disability_compensation_job.rb [hunk 5]
```diff
@@ -42,6 +44,13 @@ def perform(form526_submission_id)
       end
     end
 
+    # Fetch all claims from EVSS and return whether there are any open EP 020's.
+    # This method could be moved into a Concern when ProcessorSelector adds new job classes.
+    def pending_eps?(auth_headers)
```
Comment (yoomlam) [inline app/workers/rapid_ready_for_decision/disability_compensation_job.rb:49]: For this class, should prob move this to the `private` section.

File: app/workers/rapid_ready_for_decision/disability_compensation_job.rb [hunk 6]
```diff
@@ -42,6 +44,13 @@ def perform(form526_submission_id)
       end
     end
 
+    # Fetch all claims from EVSS and return whether there are any open EP 020's.
+    # This method could be moved into a Concern when ProcessorSelector adds new job classes.
```
Comment (yoomlam) [inline app/workers/rapid_ready_for_decision/disability_compensation_job.rb:48]: This method can go into the `Form526BaseJob` parent class.

File: app/workers/rapid_ready_for_decision/disability_compensation_job.rb [hunk 7]
```diff
@@ -42,6 +44,13 @@ def perform(form526_submission_id)
...
+    def pending_eps?(auth_headers)
+      all_claims = EVSS::ClaimsService.new(auth_headers).all_claims.body
+      all_claims['open_claims'].any? { |claim| claim['base_end_product_code'] == '020' }
```
Comment (yoomlam) [inline app/workers/rapid_ready_for_decision/disability_compensation_job.rb:51]: Checked in prod:
```ruby
s=Form526Submission.last
auth_headers=s.auth_headers
claims=EVSS::ClaimsService.new(auth_headers).all_claims.body
claims.keys
=> ["all_intent_to_files", "any_historical_claims_consolidatable", "historical_claims", "open_claims"]
opens = claims["open_claims"]
opens.first.keys
=> ["appeal_possible", "attention_needed", "base_end_product_code", "benefit_claim_type_code", 
"claim_consolidatable", "claim_phase_dates", "claim_status", "claim_tracked_items", 
"consolidated_tracked_items_list", "date", "dbq_list", "decision_notification_sent", "development_letter_sent", 
"end_product_code", "id", "max_est_claim_date", "min_est_claim_date", "new_evidence_received", 
"past_est_claim_date", "past_evidence_due_date", "past_phase_est_date", "program_type", "status", 
"status_changed", "status_type", "submitter_application_code", "submitter_role_code", "vba_document_list", 
"waiver5103_submitted", "wwd", "wwr", "wwsnfy"]

opens.any? { |claim| claim['base_end_product_code'] == '020' }
=> true

opens.select { |claim| claim['base_end_product_code'] == '020' }.size
=> 1
```

File: spec/jobs/rapid_ready_for_decision/disability_compensation_job_spec.rb [hunk 8]
```diff
@@ -119,18 +131,22 @@
...
+               auth_headers_json: auth_headers.to_json,
                saved_claim_id: saved_claim.id,
                submitted_claim_id: '600130094')
       end
 
       it 'raises an error' do
-        Sidekiq::Testing.inline! do
-          expect do
-            RapidReadyForDecision::DisabilityCompensationJob.perform_async(submission_without_account_or_edpid.id)
-          end.to raise_error RapidReadyForDecision::DisabilityCompensationJob::AccountNotFoundError
```
Comment (yoomlam) [inline spec/jobs/rapid_ready_for_decision/disability_compensation_job_spec.rb:133]: What do you think about adding this to ensure that the auth_header is what you expect:
```ruby
expect(submission_without_account_or_edpid.auth_headers['va_eauth_dodedipnid']).to be_blank
```

File: app/workers/rapid_ready_for_decision/disability_compensation_job.rb (hunk 2 above)
Comment (allthesignals) [reply app/workers/rapid_ready_for_decision/disability_compensation_job.rb:30]: Todo: change to `next`

Comment (allthesignals) [reply app/workers/rapid_ready_for_decision/disability_compensation_job.rb:30]: My understanding is that we should always be using `next` instead of `return` for these blocks to function correctly. Going to update the `return` statement a few lines down too!

File: spec/jobs/rapid_ready_for_decision/disability_compensation_job_spec.rb (hunk 8 above)
Comment (allthesignals) [reply spec/jobs/rapid_ready_for_decision/disability_compensation_job_spec.rb:133]: Sure. I do plan on refactoring this whole spec file however in a follow-up PR. But that should be a quick enough addition.

File: app/workers/rapid_ready_for_decision/disability_compensation_job.rb (hunk 2 above)
Comment (yoomlam) [reply app/workers/rapid_ready_for_decision/disability_compensation_job.rb]: We should prob add a code comment to prevent someone (like me) from changing it back to `return`.

Comment (allthesignals) [reply app/workers/rapid_ready_for_decision/disability_compensation_job.rb]: Yup that's what I meant in my previous comment

File: app/workers/rapid_ready_for_decision/form526_base_job.rb [hunk 9]
```diff
@@ -25,11 +25,21 @@ def self.rrd_claim_processed?(submission)
...
+      all_claims['open_claims'].any? { |claim| claim['base_end_product_code'] == '020' }
+    end
+
     def perform(form526_submission_id)
       form526_submission = Form526Submission.find(form526_submission_id)
 
       begin
         with_tracking(self.class.name, form526_submission.saved_claim_id, form526_submission_id) do
+          # Prefer "next" keyword here so that after-yield code continues execution
+          next if pending_eps?(form526_submission.auth_headers)
```
Comment (allthesignals) [inline app/workers/rapid_ready_for_decision/form526_base_job.rb:41]: Tests fail here because I need to wrap relevant spec in the cassette code. Can this be done in a separate PR @yoomlam?

Additionally, I'm finding I have to reference to this class method as `RapidReadyForDecision::Form526BaseJob` which... makes sense. The job method context is apparently divorced from class methods too.

Comment (yoomlam) [reply app/workers/rapid_ready_for_decision/form526_base_job.rb:41]: You'll need to get all tests passing in order to merge this PR, so you'll have to address the failing tests in this PR.

Comment (allthesignals) [reply app/workers/rapid_ready_for_decision/form526_base_job.rb:41]: @yoomlam I know that tests need to pass to merge PRs. I'm asking if moving the fix to the base class into a separate PR would make sense.

Comment (yoomlam) [reply app/workers/rapid_ready_for_decision/form526_base_job.rb]: Since it's a bug that requires a small fix, I'd apply fix ASAP. Up to you.

Comment (allthesignals) [reply app/workers/rapid_ready_for_decision/form526_base_job.rb]: @yoomlam yep, it's already applied. Can I get an approve?

File: app/workers/rapid_ready_for_decision/disability_compensation_job.rb [hunk 10]
```diff
@@ -27,9 +27,12 @@ def perform(form526_submission_id)
 
       begin
         with_tracking(self.class.name, form526_submission.saved_claim_id, form526_submission_id) do
+          # Prefer "next" keyword here so that after-yield code continues execution
```
Comment (yoomlam) [inline app/workers/rapid_ready_for_decision/disability_compensation_job.rb:30]: It's more than a preference -- it's a bug otherwise:
```suggestion
          # Use "next" keyword here so that after-yield code continues execution and job will not have a `try` status
```

File: app/workers/rapid_ready_for_decision/form526_base_job.rb [hunk 11]
```diff
@@ -25,11 +25,21 @@ def self.rrd_claim_processed?(submission)
...
+      all_claims = EVSS::ClaimsService.new(auth_headers).all_claims.body
+      all_claims['open_claims'].any? { |claim| claim['base_end_product_code'] == '020' }
+    end
+
     def perform(form526_submission_id)
       form526_submission = Form526Submission.find(form526_submission_id)
 
       begin
         with_tracking(self.class.name, form526_submission.saved_claim_id, form526_submission_id) do
+          # Prefer "next" keyword here so that after-yield code continues execution
```
Comment (yoomlam) [inline app/workers/rapid_ready_for_decision/form526_base_job.rb:40]: It's more than a preference:
```suggestion
          # Use "next" keyword here so that after-yield code continues execution and job will not have a `try` status
```

File: spec/jobs/rapid_ready_for_decision/form526_hypertension_job_spec.rb [hunk 12]
```diff
@@ -145,10 +163,12 @@
       end
 
       it 'finishes successfully' do
-        Sidekiq::Testing.inline! do
-          expect do
-            subject.perform_async(submission_without_account.id)
-          end.not_to raise_error
+        VCR.use_cassette('evss/claims/claims_without_open_compensation_claims') do
```
Comment (yoomlam) [inline spec/jobs/rapid_ready_for_decision/form526_hypertension_job_spec.rb:166]: There's a lot of repetition of this line. Wondering if there's a way to set the cassette for a `context` block.

**Review by yoomlam — CHANGES_REQUESTED** [review]

**Review by yoomlam — CHANGES_REQUESTED** [review]
Reminder to fill out the PR description.

File: app/workers/rapid_ready_for_decision/form526_base_job.rb (hunk 11 above)
Comment (allthesignals) [reply app/workers/rapid_ready_for_decision/form526_base_job.rb:40]: Will fix in a separate PR

File: app/workers/rapid_ready_for_decision/disability_compensation_job.rb (hunk 10 above)
Comment (yoomlam) [reply app/workers/rapid_ready_for_decision/disability_compensation_job.rb:30]: Here are some solutions to remedy the repetition: https://gist.github.com/myronmarston/2377461

File: spec/jobs/rapid_ready_for_decision/form526_hypertension_job_spec.rb (hunk 12 above)
Comment (yoomlam) [reply spec/jobs/rapid_ready_for_decision/form526_hypertension_job_spec.rb:166]: Here are some solutions to remedy the repetition: https://gist.github.com/myronmarston/2377461

**Review by yoomlam — DISMISSED** [review]

**Review by yoomlam — APPROVED** [review]

**Comment by allthesignals** [comment]
@rileyanderson could I get a review when you get a chance please?

**Review by allthesignals — APPROVED** [review]

**Review by yoomlam — APPROVED** [review]

**Review by thilton-oddball — APPROVED** [review]
